Core/include/Acts/Navigation/NavigationStreamHelper.hpp

```diff
@@ -0,0 +1,102 @@
+}
+
+/// Helper struct that allows to fill surfaces into the candidate vector
+///
+/// @param nStream the navigation stream that is being filled
+/// @param surfaces the surfaces that are filled in
+/// @param bTolerance the boundary tolerance used for the intersection
+/// @param abortTarget a boolean that indicates if those surfaces are abort targets
+inline static void fillSurfaces(NavigationStream& nStream,
+                                const std::vector<const Surface*>& surfaces,
+                                BoundaryTolerance bTolerance,
+                                bool abortTarget = false) {
+  std::for_each(surfaces.begin(), surfaces.end(), [&](const auto& s) {
+    nStream.candidates.push_back(NavigationStream::Candidate{
+        ObjectIntersection<Surface>(s), nullptr, bTolerance, abortTarget});
+  });
+}
+
+/// Helper struct that allows to fill portals into the candidate vector
+///
+/// @param nStream the navigation stream that is being filled
+/// @param portals the portals that are filled in
+inline static void fillPortals(NavigationStream& nStream,
+                               const std::vector<const Portal*>& portals) {
+  std::for_each(portals.begin(), portals.end(), [&](const auto& p) {
+    nStream.candidates.push_back(NavigationStream::Candidate{
+        ObjectIntersection<Surface>(&(p->surface())), p,
+        BoundaryTolerance::None()});
+  });
+}
+
+/// Initialize a stream that does not require a state object
+///
+/// @param stream [in, out] is the navigation stream to be updated
+/// @param gctx is the geometry context
+/// @param queryPoint holds current position, direction, etc.
+/// @param cTolerance is the candidate search tolerance
+///
+/// This method will first de-duplicate the candidates on basis of the surface
+/// pointer to make sure that the multi-intersections are handled correctly.
+/// This will allow intializeStream() to be called even as a re-initialization
+/// and still work correctly with at one time valid candidates.
+///
+/// @return true if the stream is active, false indicates that there are no valid candidates
+bool initializeStream(NavigationStream& stream, const GeometryContext& gctx,
+                      const NavigationStream::QueryPoint& queryPoint,
+                      BoundaryTolerance cTolerance);
+
+/// Convenience method to update a stream from a new query point,
+/// this could be called from navigation delegates that do not require
+/// a local state or from the navigator on the target stream
+///
+/// @param stream [in, out] is the navigation stream to be updated
+/// @param gctx is the geometry context
+/// @param queryPoint holds current position, direction, etc.
+///
+/// @return true if the stream is active, false indicate no valid candidates left
+bool updateStream(NavigationStream& stream, const GeometryContext& gctx,
+                  const NavigationStream::QueryPoint& queryPoint);
```

Comment:
As discussed, I wonder if this has a real application. Reintersecting and sorting is quite expensive and not as accurate as reinitialization so I would just go with reinitialization.

Reply:
Here, the navigator can use this to update the target stream (in fact, it will update only the next one, and skip over to the next but next only in case the next one is not reached).

Also, the same code can be used in the local LayerStructure stream until exhausted. That's why I put it into a helper.